### ISSUE TITLE
Bootstrap umap? 

### ISSUE THREAD

**Issue by elleobrien** [MaartenGr/BERTopic#494]
Hi, I am a mega fan of BERTopic! I notice that from run to run, the variability in cluster sizes can be a bit big(although the same dominant themes emerge, so it doesn't affect qualitative results hugely). 

I like being able to boostrap to put some upper and lower bounds on the variabilty due to sampling. It seems like you could bootstrap a few thousand UMAP manifolds (assuming compute resources are not a barrier) and then average them, and only use the average manifold for HDBSCAN. I'm curious if this has been considered, and if not, if you would have any suggestions for the right place to implement this in the code?

**Comment by MaartenGr** [comment]
Thank you for the kind words!

That is a good question. I believe that as you mentioned, compute resources would be the main bottleneck here. Although UMAP is by no means slow, running several thousand instances could take quite a while! Having said that, it might already be worthwhile to run it 10 times in order to reduce the variability that you mentioned. Do note that it might be helpful to set a `random_state` in UMAP to prevent [stochastic](https://maartengr.github.io/BERTopic/faq.html#why-are-the-results-not-consistent-between-runs) behavior. 

Took me a while to figure it out but with a little bit of hacking, you can pass UMAP embeddings to BERTopic without running it inside of BERTopic. 

First, we will need to create a class that mimics the behavior of UMAP but instead returns the embeddings as is:

```python
class EmptyUMAP:
    def fit(self, x, y):
        return self
    
    def transform(self, embeddings):
        return embeddings
```
Then, we simply pass that class to BERTopic along with the `umap_embeddings` that you bootstrapped as mentioned in your post:

```python
from bertopic import BERTopic

# This empty UMAP instance basically does nothing and allows you to pass in 
# UMAP embeddings that you generated previously
empty_umap = EmptyUMAP()
topic_model = BERTopic(umap_model=empty_umap)

# Now, we can pass the umap_embeddings and it will only apply the HDBSCAN->c-TF-IDF steps
# The embedding and dimensionality stages are ignored
topics, probs = topic_model.fit_transform(docs, umap_embeddings)
```

Hopefully, this helps a bit!

**Comment by elleobrien** [comment]
Thanks, I'm going to give this a try! Really appreciate it.

**Comment by MaartenGr** [comment]
Seeing as this was active a while ago, I am closing this for now. If you run into any issues, please let me know and I'll make sure to re-open it!

**Comment by LeoPerelli** [comment]
Hi all, I am also interested in finding a good approach to tackle the inherent UMAP variability. Did you make any relevant progress with the approach you proposed? Another option is to deal with this as kmeans does, in case a good metric to asses a dimensionality reduction exists. Running UMAP n times and just taking the run that has the best performance metric.
